### PR TITLE
add an option to set random seed for model initialization and egs shu…

### PR DESCRIPTION
**Pull request by freewym**
…ffling in training. Ref #891

### CODE REVIEW

**Comment by danpovey** [comment]
Vijay, merge this if you think it's OK.

File: egs/wsj/s5/steps/nnet3/chain/train.py [hunk 1]
```diff
@@ -127,6 +127,9 @@ def GetArgs():
 
 
     # trainer options
+    parser.add_argument("--trainer.srand", type=int, dest='srand',
+                        default = 0,
+                        help="Sets random seed for model initialization and egs shuffling")
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/chain/train.py]: ```Warning: This random seed does not control all aspects of this experiment. There might be other random seeds used other stages of the experiment like data preparation (e.g. volume perturbation).```

File: egs/wsj/s5/steps/nnet3/chain/train.py [hunk 2]
```diff
@@ -362,11 +365,11 @@ def TrainNewModels(dir, iter, num_jobs, num_archives_processed, num_archives,
   --print-interval=10 --momentum={momentum} \
   --max-param-change={max_param_change} \
    "{raw_model}" {dir}/den.fst \
-  "ark,bg:nnet3-chain-copy-egs --truncate-deriv-weights={trunc_deriv} --frame-shift={fr_shft} ark:{egs_dir}/cegs.{archive_index}.ark ark:- | nnet3-chain-shuffle-egs --buffer-size={shuffle_buffer_size} --srand={iter} ark:- ark:-| nnet3-chain-merge-egs --minibatch-size={num_chunk_per_minibatch} ark:- ark:- |" \
+  "ark,bg:nnet3-chain-copy-egs --truncate-deriv-weights={trunc_deriv} --frame-shift={fr_shft} ark:{egs_dir}/cegs.{archive_index}.ark ark:- | nnet3-chain-shuffle-egs --buffer-size={shuffle_buffer_size} --srand={srand} ark:- ark:-| nnet3-chain-merge-egs --minibatch-size={num_chunk_per_minibatch} ark:- ark:- |" \
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/chain/train.py]: Let the seed be different for each iteration. You could use `iter+srand`.

**Comment by vijayaditya** [comment]
You seem to have forgotten setting the random seed for egs shuffling jobs.

**Comment by danpovey** [comment]
Obviously it will still need to be the case that the egs-shuffling seeds
differs from iter to iter (e.g. add to it).

File: egs/wsj/s5/steps/nnet3/train_rnn.py [hunk 3]
```diff
@@ -365,11 +368,11 @@ def TrainNewModels(dir, iter, num_jobs, num_archives_processed, num_archives,
   --print-interval=10 --momentum={momentum} \
   --max-param-change={max_param_change} \
   --optimization.min-deriv-time={min_deriv_time} "{raw_model}" \
-  "ark,bg:nnet3-copy-egs {context_opts} ark:{egs_dir}/egs.{archive_index}.ark ark:- | nnet3-shuffle-egs --buffer-size={shuffle_buffer_size} --srand={iter} ark:- ark:-| nnet3-merge-egs --minibatch-size={num_chunk_per_minibatch} --measure-output-frames=false --discard-partial-minibatches=true ark:- ark:- |" \
+  "ark,bg:nnet3-copy-egs {context_opts} ark:{egs_dir}/egs.{archive_index}.ark ark:- | nnet3-shuffle-egs --buffer-size={shuffle_buffer_size} --srand={srand} ark:- ark:-| nnet3-merge-egs --minibatch-size={num_chunk_per_minibatch} --measure-output-frames=false --discard-partial-minibatches=true ark:- ark:- |" \
```
Comment (freewym) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: rand seed  for shuffling is set here.

Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: I am talking about the seed used in steps/nnet3/get_egs.sh which is used while dumping the egs.

Comment (freewym) [inline egs/wsj/s5/steps/nnet3/train_rnn.py]: Fixed/

File: egs/wsj/s5/steps/nnet3/get_egs.sh [hunk 4]
```diff
@@ -345,7 +346,7 @@ if [ $stage -le 5 ]; then
 
   if [ $archives_multiple == 1 ]; then # normal case.
     $cmd --max-jobs-run $nj JOB=1:$num_archives_intermediate $dir/log/shuffle.JOB.log \
-      nnet3-shuffle-egs --srand=JOB "ark:cat $egs_list|" ark:$dir/egs.JOB.ark  || exit 1;
+      nnet3-shuffle-egs --srand=$[JOB+$srand] "ark:cat $egs_list|" ark:$dir/egs.JOB.ark  || exit 1;
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/get_egs.sh]: I don't think command will work. bash would replace $[JOB+$srand] with the value of srand.

**Comment by vijayaditya** [comment]
don't forget changes to `steps/nnet3/chain/get_egs.sh`

**Comment by danpovey** [comment]
It *might* work if the first $ is escaped, like
 \$[JOB+$srand]

Dan

**Comment by freewym** [comment]
Yes. escape of $ works.

**Comment by vijayaditya** [comment]
Remember to store the random seed used for an experiment in the experiment directory (e.g. `exp/nnet3/tdnn`. You could also add checks 

1. to see if the random seed used for generating the egs is the same as the one used for model initialization and egs shuffling in the training script
2. to verify that the random seed being used when restarting the training from an intermediate stage is the same as the one stored in the experiment directory.

When either of these checks fail please print a warning and report which random seed is being used.

@danpovey  I was wondering if we should have a random seed associated with an experiment (which includes data-preparation, egs dumping and training). We could store this `random_seed` file in the s5/config directory. We could ensure all scripts in `local/` access this file and supply it to the appropriate `steps/*` scripts via command-line. It is a bit tricky to ensure that the same random seed is used consistently as we usually reuse feature files or egs across experiments. Any comments ?

**Comment by freewym** [comment]
Added the random seed checks.

**Comment by danpovey** [comment]
Vijay, regarding the random seed, I don't think storing it in s5/config is a good idea because typically the only reason you'd want to do this is if you were doing multiple runs with the same experiment, and they'd all be in the same s5 directory.  I think it's best to put it inside a script and pass it as appropriate to scripts that that script calls.  And I don't think it's necessary to go overboard with the random seed and use it for things feature generation.  This will make next to no difference and just provides a lot of clutter.  There are other things that can affect the results, including the hardware it's run on, and non-reproducible differences caused by non-determinism in CUBLAS.  In any case, it will probably be rare that we'll want to run with different random seeds.

Actually I would have been happy with just setting the random seed in the neural net initialization in nnet-init.  I suspect you get 90% of the difference from that.
